Profile scrollers

emalysz opened the pull request.

Commit:
First attempt: sliders to dictate range, grey boxes in separate canvas, no conversions.

The markers have now been added to a canvas to the right of the main canvas.
The two arrows (top and bottom ranges) scroll throughout the entire range.
Refocus filters the data to the specified range.
There are issues with the fact that we are overlaying a canvas.
We need to include time conversions in order to go from absolute, relative,
and screen locations. This is a WIP

squarewave commented:
Okay, I'm going to start off with some broad comments here. I may dive into the details, but I think some conceptual reorganizing is a first priority.

First off, we should avoid creating a canvas that gets overlaid on another canvas. I'm not sure what optimizations we've made to limit what we actually have to draw with a big canvas, but my understanding is we basically have to process every pixel inside a canvas, at least when it comes to the compositor. It's not the end of the world, but if it's easy to avoid, we should. (And it is easy to avoid here.) So let's just use the main renderer, and add a rect for each overlay region that's partially filled:

```
let {
  tracks,
} = gState;
let {
  constrainedTimeBegin,
} = gOuterState;
if (isTopSlider) {
  renderer.pushRect(TIMELINE_UNSELECTED_GREY,
                    /*x:*/ 0, /*y:*/ 0,
                    /*width:*/ tracks.length, /*height:*/ absTimeToRelTime(constrainedTimeBegin),
                    TIMELINE_SELECTION_OVERLAY_DEPTH,
                    TIMELINE_SELECTION_OVERLAY_FILL);
} else {
  ...
}
```

There's a few things introduced in the snippet above though that I should call out. The first is that I made a `gOuterState` global for holding the state of our scrollers. This will make it easier to track and even reset them if we'd like to, and makes it clearer when reading the code that these are not local variables. I didn't wrap the state into the `gState` global, because in a sense this supersedes that. `gState` should only contain information from the range that hasn't been filtered out by the data in `gOuterState`.

The second thing is I invoke a function here which doesn't exist: `absTimeToRelTime`. Currently I do all conversion between coordinate systems inline, and it's a bit hard to follow. So we should make these more explicit. There are three Y coordinate systems that we have to deal with: absolute time (seconds since epoch), relative time (seconds since first entry in filtered data), and screen space (pixels from the top of the page). In general, I like to stay in time coordinates, and relative time coordinates where possible. This is because it's the functional space we're working in, rather than just the presentational space. In our case, we will want to store our filter range in absolute time. We can't do relative time, because in the case where we've filtered our entries down to a particular range, relative time is actually *determined* by the start of that range. I will just include all the coordinate conversion functions below. We should replace all of our coordinate conversion code with these functions. (I can do this as a followup, if you like.)

```
function absTimeToRelTime(absTime) {
  let { minTime } = gState;
  return absTime - minTime;
}

function relTimeToAbsTime(relTime) {
  let { minTime } = gState;
  return relTime + minTime;
}

function relTimeToScreenSpace(relTime) {
  let { rendererScale, rendererTranslate } = gState;
  return (relTime + rendererTranslate) * rendererScale;
}

function absTimeToScreenSpace(absTime) {
  return relTimeToScreenSpace(absTimeToRelTime(absTime));
}

function screenSpaceToRelTime(screenY) {
  let { rendererScale, rendererTranslate } = gState;
  return screenY / rendererScale - rendererTranslate;
}

function screenSpaceToAbsTime(screenY) {
  return relTimeToAbsTime(screenSpaceToRelTime(screenY));
}
```

So, in order to compute our `constrainedTimeBegin` value from a mouse coordinate, we would just do `gOuterState.constrainedTimeBegin = screenSpaceToAbsTime(mouseY);`. In order to filter our data to only things within the constrained time, we could do something like `let constrainedData = gOuterState.unconstrainedData.filter(row => row.start > gOuterState.constrainedTimeBegin && row.start + row.duration < gOuterState.constrainedTimeEnd);`. Does that all seem reasonable, and fit with what you've encountered so far in developing this?

Commit:
This will enable a user to specify a certain range within the procmon tool.

After review feedback 1, I implemented time conversions, removed overlay canvas,
and changed the implementation of the grey out selection rectanges.

To mimic the behavior of the firefox profiler, the user will not be able to scroll
or zoom once a range has been focused. They can use these features without a focused
range, and the range markers will disappear.

When a user focuses a range, additional timeline markers will indicate the top and bottom
range numbers.

emalysz commented on procmonAnalyze.js:
```diff
@@ -331,48 +358,72 @@ async function drawData(data, diskify) {
 }
 
-function drawBackground() {
+function drawBackground(isRefocused) {
```
@squarewave So the problem I'm facing is now with `drawBackground` once we've scrolled. It seems like things are working when we create top/bottom constraints (correct foreground showing, correct timeline indicators, etc). The problem is when we scroll from a specific section. I feel like the problem has something to do with `let offsetPx = (offset + targetRendererTranslate) * rendererScale;`. We would want the timeline indicators to change depending on our scroll, and right now they just push the original numbers up. Do you have any suggestions as to how I can fix this?

Draw foreground seems to be working fine with the scroll, it just has an incorrect background associated with it.

Commit:
general code cleanup and nit fixes

squarewave commented on index.html:
```diff
@@ -22,6 +22,7 @@
   </div>
 
   <canvas id="canvas"></canvas>
+  <canvas id="canvas-markers"></canvas>
```
I think this was in a previous comment, but maybe not? Apologies if I didn't catch it last time. I don't think the markers should be a whole canvas. Just HTML+CSS should be simpler. Something like:

```css
.left-arrow {
  width: 0; 
  height: 0; 
  border-top: 10px solid transparent;
  border-bottom: 10px solid transparent; 
  border-right:10px solid black; 
}
```

squarewave commented on procmonAnalyze.js:
```diff
@@ -15,16 +15,34 @@ const HOVERED_ENTRY_FILL = 0.9;
+var data, diskifyData; // Contains the entire array of data from input file
+
+var offsetX, offsetY;
+reOffset();
```
I think we can get rid of this if we have HTML/CSS-based markers, yes?

squarewave commented on procmonAnalyze.js:
```diff
@@ -15,16 +15,34 @@ const HOVERED_ENTRY_FILL = 0.9;
+
+var offsetX, offsetY;
+reOffset();
+window.onscroll=function(e){ reOffset(); }
```
Hmm. I would rather we make sure that the window doesn't scroll, rather than add a scroll handler. Thoughts?

squarewave commented on procmonAnalyze.js:
```diff
@@ -15,16 +15,34 @@ const HOVERED_ENTRY_FILL = 0.9;
+var offsetX, offsetY;
+reOffset();
+window.onscroll=function(e){ reOffset(); }
+window.onresize=function(e){ reOffset(); }
```
Let's put this at the bottom with all of the other listeners, and change it to `window.addEventListener("resize", () => reOffset());`. Also, I could you rename it to something like `recomputeRefocusMarkerOffsets()`?

squarewave commented on procmonAnalyze.js:
```diff
@@ -15,16 +15,34 @@ const HOVERED_ENTRY_FILL = 0.9;
 const canvas = document.getElementById("canvas");
 const fsmapCanvas = document.getElementById("fsmap-canvas");
+const markersCanvas = document.getElementById("canvas-markers");
+var data, diskifyData; // Contains the entire array of data from input file
```
Here and below: can you move all global state under `gOuterState`, rather than `var`s at the root scope?

squarewave commented on procmonAnalyze.js:
```diff
@@ -184,7 +206,7 @@ function parseReadDetail(detail) {
 }
 
-async function drawData(data, diskify) {
+async function drawData(aData, diskify, isRefocused) {
```
Let's not change arguments to hungarian notation (`aData`) here - moving `data` under `gOuterState` should be sufficient.

squarewave replied:
Also, I don't think you need `isRefocused` here? Just set `gOuterState.data` to the value you pass into `drawData` from `readFileContents`.

squarewave commented on procmonAnalyze.js:
```diff
@@ -334,65 +356,102 @@ async function drawData(data, diskify) {
-    lastHoveredRect: null,
-    selectedEntry: null,
-    activePath: null,
+      minTime,
```
I believe your editor reindented this to 4 spaces. Could you bring it back down to 2?

squarewave commented on procmonAnalyze.js:
```diff
@@ -334,65 +356,102 @@ async function drawData(data, diskify) {
+      activePath: null,
+      topMarkerCoordinate: 0,
+      bottomMarkerCoordinate: markersCanvas.height,
+      isRefocused: isRefocused,
```
Do we need `isRefocused` as separate state, when it can always be recomputed as `gOuterState.minTime != gOuterState.constrainedTimeBegin`? We could add a helper method `function isRefocused() {...}` to help make it cleaner to read when we need it.

squarewave commented on procmonAnalyze.js:
```diff
@@ -334,65 +356,102 @@ async function drawData(data, diskify) {
+      topMarkerCoordinate: 0,
+      bottomMarkerCoordinate: markersCanvas.height,
+      isRefocused: isRefocused,
+      minTimeRelative: absTimeToRelTime(minTime),
+      maxTimeRelative: absTimeToRelTime(maxTime),
```
Similar to the above: do we need these as separate state, when they can always be computed from `minTime`? I just don't want to add more state which can get out of sync.

squarewave commented on procmonAnalyze.js:
```diff
@@ -334,65 +356,102 @@ async function drawData(data, diskify) {
+      maxTimeRelative: absTimeToRelTime(maxTime),
   };
 
+  gOuterState.constrainedTimeBegin = minTime;
```
Let's move this outside of `drawData` calls. I think it would be preferable if this function does not modify `gOuterState`, and as much as possible restricts itself to thinking about only `gState`.

squarewave commented on procmonAnalyze.js:
```diff
@@ -413,33 +472,73 @@ function drawBackground() {
+  drawRangeControl(bottomRange, false);
+
+
+  markerCtx.onmousedown=(function(e){handleMouseDown(e);});
+  markerCtx.onmousemove=(function(e){handleMouseMove(e);});
+  markerCtx.onmouseup=(function(e){handleMouseUpOut(e);});
+  markerCtx.onmouseout=(function(e){handleMouseUpOut(e);});
```
I think this is a bit problematic... registering these handlers here conflicts with the registrations on the `document` below. Can we just integrate the marker handling logic into the existing mouse handlers? I think the way it's set up here also causes problems because if you mouse out of the marker canvas, it will stop adjusting. You should be able to click and drag anywhere and have it move the marker just according to your mouse's Y position.

squarewave replied:
Also, as a general note: I would keep the registration of event handlers outside of any `draw...` functions. It's a bit of a misleading name, and you end up setting the event handlers more than once, which is less than ideal.

squarewave commented on procmonAnalyze.js:
```diff
@@ -1429,12 +1655,78 @@ function handleMouseUp(event) {
+  }
+}
+
+function absTimeToRelTime(absTime) {
```
Relative time, as the code has been informally treating it prior to this, was always relative to the `minTime` as defined by `gState`. If we use `gOuterState` here, then we're going to run into problems, and some of the bugs I'm running into feel like that's what's going on?

squarewave commented on procmonAnalyze.js:
```diff
@@ -474,17 +577,26 @@ function drawForeground() {
-      let endRelative = entry.end - minTime;
-      let startPixels = (startRelative + rendererTranslate) * rendererScale;
-      let endPixels = (endRelative + rendererTranslate) * rendererScale;
+      let startRelative = absTimeToRelTime(entry.start);
```
Here is an example of the problem with `absTimeToRelTime` - we're subtracting `constrainedRelative` below to try to correct for the discrepancy, but we shouldn't need to do that. The renderer should be entirely in relative time, relative to `gState.minTime`, not `gOuterState.minTime`, because it shouldn't need to be aware of `gOuterState`. There are some leaky parts, like wanting to make sure that the timeline indicators and things like that use time relative to `gOuterState`, but I think that means we need another coordinate space, maybe "outer relative time" (`absTimeToOuterRelTime`). Is that making sense? (I could be wrong about this, but I think in the vast majority of cases things are simpler if we have relative time be relative to `gState.minTime`.)

squarewave reviewed (commented):
I haven't gone through everything yet, and I intend to go through more on (hopefully) Monday, but I will leave these comments here since I think there's a good chunk of things to be done addressing them.

One of the comments alluded to a bug I noticed, which I'll give STR for here:

1. Open a procmon log in the viewer
2. Refocus so the top is around, say, 5 seconds, and the bottom is around 10 seconds.
3. Keep refocusing the top to somewhere between 5 and 6 seconds
4. Notice that eventually, the top somehow actually refocuses past 6 seconds
